sklearn/utils/metadata_requests.py

```diff
@@ -0,0 +1,748 @@
+        looks for the information available in the set default values which are
+        set using ``_metadata_request__*`` class attributes.
+
+        References
+        ----------
+        .. [1] https://www.python.org/dev/peps/pep-0487
+        """
+        try:
+            requests = cls._get_default_requests().to_dict()
+        except Exception:
+            # if there are any issues in the default values, it will be raised
+            # when ``get_metadata_request`` is called. Here we are going to
+            # ignore all the issues such as bad defaults etc.`
+            super().__init_subclass__(**kwargs)
+            return
+
+        for request_method, request_keys in requests.items():
+            # set ``{method}_requests``` methods
+            if not len(request_keys):
+                continue
+            setattr(
+                cls,
+                f"{request_method}_requests",
+                RequestMethod(request_method, sorted(request_keys)),
+            )
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def _get_default_requests(cls):
+        """Collect default request values.
+
+        This method combines the information present in ``metadata_request__*``
+        class attributes.
+        """
+        requests = MetadataRequest()
+
+        # need to go through the MRO since this is a class attribute and
+        # ``vars`` doesn't report the parent class attributes. We go through
+        # the reverse of the MRO since cls is the first in the tuple and object
+        # is the last.
+        defaults = defaultdict()
+        for klass in reversed(inspect.getmro(cls)):
+            klass_defaults = {
+                attr: value
+                for attr, value in vars(klass).items()
+                if attr.startswith("_metadata_request__")
+            }
+            defaults.update(klass_defaults)
+        defaults = dict(sorted(defaults.items()))
+
+        # First take all arguments from the method signatures and have them ass
+        # ERROR_IF_PASSED, except X, y, *args, and **kwargs.
+        for method in METHODS:
+            # Here we use `isfunction` instead of `ismethod` because calling `getattr`
+            # on a class instead of an instance returns an unbound function.
+            if not hasattr(cls, method) or not inspect.isfunction(getattr(cls, method)):
+                continue
+            for pname, param in inspect.signature(
+                getattr(cls, method)
+            ).parameters.items():
+                if pname in {"X", "y", "self", "cls"}:
+                    continue
+                if param.kind in {param.VAR_POSITIONAL, param.VAR_KEYWORD}:
+                    continue
+                getattr(requests, method).add_request(
+                    prop=pname,
+                    alias=RequestType.ERROR_IF_PASSED,
+                    allow_aliasing=False,
+                    overwrite=False,
+                )
```

Comment:
The tests fail because now we're doing signature check, and all splitters accept `groups`, and the ones which don't actually need `groups` have the default request set as `ERROR_IF_PASSED`, which results in an error in some tests. We have a few options here:

- only splitters which need groups to have it as a kwarg.
- have a way to override the signature check and remove it if explicitly set (somehow through `_metadata_request__groups`

Also, is there a way for me not to have `"self"` and `"cls"` hard-coded here?

Reply:
> have a way to override the signature check and remove it if explicitly set (somehow through _metadata_request__groups

Yes, this.

> Also, is there a way for me not to have "self" and "cls" hard-coded here?

If inspect.ismethoddescriptor you can ignore the first arg?